### PR TITLE
[Transform] Module Interface for Transform

### PR DESCRIPTION
**Pull request by mufeili**
## Description
- This PR adds module interfaces for a subset of transform functions and extends them for simplicity and better user experience, e.g., support for heterogeneous graphs.
- The design is motivated by the [transform modules in PyTorch Geometric](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html).
- Compared with functional interfaces, the module interfaces provide a better experience for applying a same transform with customizable arguments to multiple graphs.
- The module interfaces have also been utilized in Torch Vision for data augmentation.
- The PR also refactors the previous functional interfaces for transform functions.
- Future PRs will add more transform modules for data augmentation and transformation.

@sojiadeshina @zjost It will be great if you can take a look and see if there are growth opportunities for user experience. Thanks a lot!

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: docs/source/api/python/dgl.rst [hunk 1]
```diff
@@ -48,40 +48,6 @@ Operators for extracting and returning subgraphs.
...
-    to_homogeneous
-    to_networkx
-    line_graph
-    khop_graph
-    metapath_reachable_graph
-    adj_product_graph
-    adj_sum_graph
-    reorder_graph
-    sort_csr_by_tag
-    sort_csc_by_tag
```
Comment (jermainewang) [inline docs/source/api/python/dgl.rst:83]: does it mean they will not be available as `dgl.xxx`?

File: python/dgl/transform/module.py [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+                new_g.nodes[ntype].data[key] = feat
+
+        if copy_edata:
+            for c_etype in g.canonical_etypes:
+                for key, feat in g.edges[c_etype].data.items():
+                    new_g.edges[c_etype].data[key] = feat
+
+        return new_g
+
+    def __call__(self, g):
```
Comment (jermainewang) [inline python/dgl/transform/module.py:131]: I feel we only need `__call__` and `__repr__`. Could you explain why we need the other methods?

**Comment by mufeili** [comment]
@jermainewang 

1. I've restored the functional transform APIs in the documentation page for `dgl`.
2. The module interfaces are listed in the documentation page for `dgl.transform`.
3. I've removed the functions of `BaseTransform` other than `__call__` and `__repr__`.
4. There are some additional issues spotted in the unit tests for modules depending on the functional APIs.
    - `khop_graph` always return a `DGLGraph` of idtype int64 on GPU, regardless of the idtype and device of the input graph. We can fix it by always returning a graph of the same idtype and device as the input graph. This will be a breaking change though.
    - `to_simple` only support graphs on CPU. A quick fix can be moving the graph to CPU first and then moving the transformed graph to GPU. Should we implement that at the level of module interfaces or functional interfaces?

File: python/dgl/transform/functional.py [hunk 3]
```diff
@@ -13,30 +13,30 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 #
-"""Module for graph transformation utilities."""
+"""Functional interface for"""
```
Comment (jermainewang) [inline python/dgl/transform/functional.py:16]: incomplete description here

File: python/dgl/transform/module.py [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+    Description
+    -----------
+    Add self-loops for each node in the graph and return a new graph.
+
+    For heterogeneous graphs, self-loops are added only for edge types with same
+    source and destination node types.
+
+    Parameters
+    ----------
+    remove_first : bool, optional
```
Comment (jermainewang) [inline python/dgl/transform/module.py:103]: `allow_duplicate=False`

File: python/dgl/transform/module.py [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+        If True, it will first remove self-loops to prevent duplicate self-loops.
+    add_self_etypes : bool, optional
```
Comment (jermainewang) [inline python/dgl/transform/module.py:105]: `new_etype=False`

File: python/dgl/transform/module.py [hunk 6]
```diff
@@ -0,0 +1,790 @@
...
+
+    For a heterogeneous graph, it adds a "reverse" edge type for each edge type
+    to hold the reverse edges. For example, for a canonical edge type ('A', 'r', 'B'),
+    it adds a canonical edge type ('B', 'rev_r', 'A').
+
+    Parameters
+    ----------
+    copy_edata : bool, optional
+        If True, the features of the reverse edges will be identical to the original ones.
+    combine_like : bool, optional
```
Comment (jermainewang) [inline python/dgl/transform/module.py:283]: `new_etype=False`

File: python/dgl/transform/module.py [hunk 7]
```diff
@@ -0,0 +1,790 @@
...
+
+    Description
+    -----------
+    Convert a graph to a simple graph without parallel edges and return a new graph.
+
+    Parameters
+    ----------
+    return_counts : str, optional
+        The edge feature name to hold the edge count in the original graph.
+    writeback_mapping : bool, optional
```
Comment (jermainewang) [inline python/dgl/transform/module.py:437]: This shouldn't be allowed right? It cannot be used together with `Compose`.

File: python/dgl/transform/module.py [hunk 8]
```diff
@@ -0,0 +1,790 @@
...
+    of edges :math:`(u, v)` and :math:`(v, w)` in :math:`G`, there will be an
+    edge from the node corresponding to :math:`(u, v)` to the node corresponding to
+    :math:`(v, w)` in :math:`L(G)`.
+
+    This module only works for homogeneous graphs.
+
+    Parameters
+    ----------
+    backtracking : bool, optional
+        If True, the line graph node corresponding to :math:`(u, v)` will not have an
```
Comment (jermainewang) [inline python/dgl/transform/module.py:525]: `If False, ...`?

File: python/dgl/transform/module.py [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+
+    def __call__(self, g):
+        return functional.khop_graph(g, self.k)
+
+class AddMetaPaths(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Add canonical edge types to an input graph based on given metapaths, as described in
```
Comment (jermainewang) [inline python/dgl/transform/module.py:601]: `Add new edges to an input graph based on given metapaths, ...`

File: python/dgl/transform/module.py [hunk 10]
```diff
@@ -0,0 +1,790 @@
...
+
+    The following examples use PyTorch backend.
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import KNNGraph
+
+    >>> g = dgl.rand_graph(5, 20)
+    >>> g.ndata['h'] = torch.randn(g.num_nodes(), 3)
+    >>> transform = KNNGraph(ndata_name='h', k=3)
```
Comment (jermainewang) [inline python/dgl/transform/module.py:738]: This is a little bit strange. Shouldn't it transform a tensor to a graph?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/transform/module.py (hunk 10 above)
Comment (mufeili) [reply python/dgl/transform/module.py:738]: In that case, it cannot be used in `Compose` as well.

File: python/dgl/transform/module.py (hunk 5 above)
Comment (mufeili) [reply python/dgl/transform/module.py:105]: Then will this still add reverse canonical edge types for asymmetric canonical edge types?

File: python/dgl/transform/module.py (hunk 10 above)
Comment (jermainewang) [reply python/dgl/transform/module.py:738]: True. Then why should we have this? How people use it in real cases?

File: python/dgl/transform/module.py (hunk 5 above)
Comment (jermainewang) [reply python/dgl/transform/module.py:105]: It will.

Comment (jermainewang) [reply python/dgl/transform/module.py:105]: You could change it to a more explicit name.

**Comment by mufeili** [comment]
> @jermainewang
> 
> 1. I've restored the functional transform APIs in the documentation page for `dgl`.
> 2. The module interfaces are listed in the documentation page for `dgl.transform`.
> 3. I've removed the functions of `BaseTransform` other than `__call__` and `__repr__`.
> 4. There are some additional issues spotted in the unit tests for modules depending on the functional APIs.
>    
>    * `khop_graph` always return a `DGLGraph` of idtype int64 on GPU, regardless of the idtype and device of the input graph. We can fix it by always returning a graph of the same idtype and device as the input graph. This will be a breaking change though.
>    * `to_simple` only support graphs on CPU. A quick fix can be moving the graph to CPU first and then moving the transformed graph to GPU. Should we implement that at the level of module interfaces or functional interfaces?

* `khop_graph` now returns a `DGLGraph` whose idtype and device are as the same as the input graph.
* `to_simple` still only supports CPU. I've disabled the unit test for GPU.

File: python/dgl/transform/module.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Changed.

File: python/dgl/transform/module.py (hunk 5 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Changed.

File: python/dgl/transform/module.py (hunk 6 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Changed.

File: python/dgl/transform/module.py (hunk 7 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Removed.

File: python/dgl/transform/module.py (hunk 8 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Changed.

File: python/dgl/transform/module.py (hunk 9 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Changed.

File: python/dgl/transform/module.py (hunk 10 above)
Comment (mufeili) [reply python/dgl/transform/module.py]: Removed this module.

**Review by jermainewang — APPROVED** [review]
LGTM

**Pull request merged**